examples/cluster/plot_kmeans_assumptions.py

```diff
@@ -3,67 +3,169 @@
-)
-y_pred = KMeans(n_clusters=3, n_init="auto", random_state=random_state).fit_predict(
-    X_varied
-)
+axs[0, 1].scatter(X_aniso[:, 0], X_aniso[:, 1], c=y)
+axs[0, 1].set_title("Anisotropically Distributed Blobs")
+
+axs[1, 0].scatter(X_varied[:, 0], X_varied[:, 1], c=y_varied)
+axs[1, 0].set_title("Unequal Variance")
 
-plt.subplot(223)
-plt.scatter(X_varied[:, 0], X_varied[:, 1], c=y_pred)
-plt.title("Unequal Variance")
+axs[1, 1].scatter(X_filtered[:, 0], X_filtered[:, 1], c=y_filtered)
+axs[1, 1].set_title("Unevenly Sized Blobs")
+
+plt.suptitle("Ground truth clusters").set_y(0.95)
+plt.show()
+
+# %%
+# Fit models and plot results
+# ---------------------------
+#
+# The previously generated data is now used to show how
+# :class:`~sklearn.cluster.KMeans` behaves in the following scenarios:
+#
+# - Non-optimal number of clusters: in a real setting there is no uniquely
+#   defined **true** number of clusters. An appropriate number of clusters has
+#   to be decided from data-based criteria and knowledge of aim.
+# - Anisotropically distributed blobs: k-means consists of minimizing sample's
+#   euclidean distances to the centroid of the cluster they are assigned to. As
+#   a consequence, k-means is more appropriate for clusters that are isotropic
+#   and normally distributed (i.e. spherical gaussians).
+# - Unequal variance: k-means is equivalent to taking the maximum likelihood
+#   estimator for a "mixture" of k gaussian distributions with the same
+#   variances but with possibly different means.
+# - Unevenly sized blobs: there is no theoretical result about k-means that
+#   states that it requires similar cluster sizes to perform well, yet
+#   minimizing euclidean distances does mean that the more sparse and
+#   high-dimensional the problem is, the higher is the need to run the algorithm
+#   with different centroid seeds to ensure a global minimal inertia.
+
+from sklearn.cluster import KMeans
+
+common_params = {
+    "n_init": "auto",
+    "random_state": random_state,
+}
+
+fig, axs = plt.subplots(nrows=2, ncols=2, figsize=(12, 12))
+
+y_pred = KMeans(n_clusters=2, **common_params).fit_predict(X)
+axs[0, 0].scatter(X[:, 0], X[:, 1], c=y_pred)
+axs[0, 0].set_title("Non-optimal Number of Clusters")
+
+y_pred = KMeans(n_clusters=3, **common_params).fit_predict(X_aniso)
+axs[0, 1].scatter(X_aniso[:, 0], X_aniso[:, 1], c=y_pred)
+axs[0, 1].set_title("Anisotropically Distributed Blobs")
+
+y_pred = KMeans(n_clusters=3, **common_params).fit_predict(X_varied)
+axs[1, 0].scatter(X_varied[:, 0], X_varied[:, 1], c=y_pred)
+axs[1, 0].set_title("Unequal Variance")
+
+y_pred = KMeans(n_clusters=3, **common_params).fit_predict(X_filtered)
+axs[1, 1].scatter(X_filtered[:, 0], X_filtered[:, 1], c=y_pred)
+axs[1, 1].set_title("Unevenly Sized Blobs")
+
+plt.suptitle("Unexpected KMeans clusters").set_y(0.95)
+plt.show()
+
+# %%
+# Possible solution
+# -----------------
+#
+# For an example on how to find a correct number of blobs, see
```

Comment:
Can you make the text describing the different solutions have the same order as the code examples below? It is possible to work out which bit of text is meant to go with which code snippet but I think it would be easier if the first solution described in the text was also the first code snippet, and so on.

Reply:
Also sounds good to me